### ISSUE TITLE
when forwarding mail, only first attachment is sent, all subsequent ones are 0 bytes

### ISSUE THREAD

**Issue by makinghappen** [nextcloud/mail#6064]
### Steps to reproduce

1. send an email with several attachments(2 or more) from nextcloud mail app
2. go to send folder and forward it again
3. only first attachment is sent now, all subsequent ones are 0 bytes

### Expected behavior

all attachments must be forwarded

### Actual behavior

second and subsequent attachments are not sent, empty file is sent instead:
Content-Type: application/x-empty; name=photo.jpg

Content-Disposition: attachment; filename=photo.jpg

Content-Transfer-Encoding: base64

### Mail app version

1.11.7

### Operating system

Debian

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database

MySQL

### Additional info

when forwarding from mail client everything works.

**Comment by makinghappen** [comment]
any hope to fix soon?

**Comment by ChristophWurst** [comment]
@GretaD you +1'd. Are you able to reproduce?

**Comment by boreevyuri** [comment]
I think it's bug in buildAttachmentsPartsQuery in [buildAttachmentsPartsQuery](https://github.com/nextcloud/mail/blob/623018da8bdab66b4359f5f5c91f5d26ac47b937/lib/IMAP/MessageMapper.php#L613)

When getRawAttachments tries to get second attachment from source letter, buildAttachmentsPartsQuery returns array with empty first element (or several preceding empty elements if letter has more than two attachments).

Temporary fixed this on own installation with `if (!empty($decoded)) { $attachments[] = $decoded; }` in getRawAttachments but it looks like something wrong.

**Comment by miaulalala** [comment]
Very interesting, thanks for taking a look!

**Comment by sazanof** [comment]
I faced this problem from the other side. I opened the email containing the attachment and tried to forward it.
After I found out that there were no attachments at all in the forwarded email.

I tried to fix the frontend, but it looks like this alone will not do. Through the "corrected frontend" I received a visualization of email attachments (forwarded email). Next - I forward them and lo and behold - the letter arrives and there are attachments in it! Only something is wrong: the icons are not the same. The pdf has an archive icon. Аnd the `json` containing the `attachments->attachment` object has an invalid mime type

I decided to dig deeper ... As a result, what I managed to dig up : =)

1) by clicking on send a message, an array with attachments is sent (correct)
1.1) So far, the type is sewn tightly here https://github.com/nextcloud/mail/issues/6227 , I have temporarily commented out these lines and made a dirty fix
https://github.com/sazanof/nextcloud_mail/blob/bug/no-attachments-when-forwarding/lib/Controller/OutboxController.php#L122 (maybe it's creepy, but let me continue the story =))
2) the `handleAttachments` method is called on the server, which at one time pulls the `handleForwardedAttachment` method
3) in this method, in the line https://github.com/nextcloud/mail/blob/main/lib/Service/Attachment/AttachmentService.php#L314
the $attachmentMessage message is incorrectly located, as it searches by uid (this is our `(int)$attachment['messageId']`)
All due to the fact that the `uid` is not unique for the user. and it is unique only for `mailboxId`+`uid`
4) as a result, I realized that the attachment (attachments) are attached **from another letter**!

_In general, as a temporary measure, I shoved mailboxId into each attachment in the getBody method and while my crutches are working ... =)_

actually, it may be useful at least a little... - https://github.com/sazanof/nextcloud_mail/commit/e7eb3ce07b602d59e983e6a5270f5d8294ae00c9

Can I add the necessary fields (mailboxId at least) to the local message table?

I don't know what this monologue is about - I don't know if you already have ways to solve the problem, whether you are aware of the situation or not. And I would not really like to go separately with your approach.

I could organize a normal pullrequest, but maybe we need to discuss it first...

Thank you for finishing reading! =)

**Comment by miaulalala** [comment]
> the $attachmentMessage message is incorrectly located, as it searches by uid (this is our (int)$attachment['messageId'])
All due to the fact that the uid is not unique for the user. and it is unique only for mailboxId+uid

I'm not sure that this is actually what's happening.
The DB query does indeed look for the primary ID in the MessageMapper:
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Db/MessageMapper.php#L124-L136
The `$uid` is (confusingly named) the user id.

What could be happening is that the frontend is sending the wrong ID for the attachment. Can you take a look at the request to see if your 'messageId'' value is actually valid?

Thanks :)

**Comment by sazanof** [comment]
> Thanks :)

Yes, everything is correct in the `findByUserId` function.

So, I tried to reproduce the problem again and that I understood (at least tried)
==FORWARDING A MESSAGE WITH AN ATTACHMENT==

0) Parsing the url **/apps/mail/box/16/thread/9541**
1) Open the message with the attachment in the branch and click the "forward" button. (the `getBody(9541)` method of the `MessagesController` is called)
2) The `$id` parameter of getBody(9541) is the `id` identifier in the mail_messages table (**remember this**)
2.1) Received the correct message (`$this->mailManager->GetMessage` and so on). Now this method is looking for attachments...
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Controller/MessagesController.php#L238
3) To get the attachments property, in the `getBody` method pulls `$json = $this->mailManager->getImapMessage`
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Controller/MessagesController.php#L245
3.1) Oh yes, we received all the correct data and displayed them on the page. If you look at the **attachments** object, then each element of it will contain a messageId according to the uid from the mail_messages table, in a row with id **9541** (for example, messageId: **129**).
![error-forward](https://user-images.githubusercontent.com/3595562/165684888-f544a775-c943-4998-a623-2d3894b3bbef.png)
4)... here it is necessary to correct the backend so that these attachments are located, so for now you clearly have the 'local' type sewn up. let's say it's done.
5) OutboxController -> create is called(.,.,.,.,.,.,.,$ attachments)
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Controller/OutboxController.php#L100
5.1) The attachments object is identical to the `$attachments` object that we received earlier in 3.1
5.2) `$this->service->saveMessage($account, $message, $to, $cc, $bcc, $attachments);`pulls `handleAttachments`, which in our case pulls `handleForwardedAttachment`
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Controller/OutboxController.php#L129
5.2) _Hooray, we've reached the most interesting part, Sherlock._
5.3) `$attachmentMessage = $this->mailManager->getMessage($account->getUserId(), (int)$attachment['messageId']);`. The parameter `$attachment['messageId']` in our example is **129**
https://github.com/nextcloud/mail/blob/ef0c0054dca182f4a0f82514f42a9800035a1dac/lib/Service/Attachment/AttachmentService.php#L314
6) Here is the most interesting. Look at point **2)**. It turns out that we are looking for a message not with the identifier **9541**, but with the identifier 129. Therefore, the wrong message is returned from the database (or nothing at all). Therefore, further ...
7)
```
$attachments = $this->messageMapper->getRawAttachments(
			$client,
			$mailbox->getName(),
			$attachmentMessage->getUid(),
			[
				$attachment['id'] ?? []
			]
		);
```
the wrong data is being transmitted ($mailbox->getName() from `mail_messages->id->129`)
and accordingly, the attachments are returned the wrong ones, or they are not returned at all
I hope I didn't bore or confuse you too much =). Have a nice day!

**Comment by kesselb** [comment]
Sounds similar to https://github.com/nextcloud/mail/pull/5602.

**Comment by st3iny** [comment]
Anna and I did an investigation:

## 1. 

We set `$attachment['messageId']` to the uid of the message.

https://github.com/nextcloud/mail/blob/a96e8d5eaa6c9f44a175f09c571f157283cd9ec7/lib/Model/IMAPMessage.php#L404


However, our code expects it to be the database id.

https://github.com/nextcloud/mail/blob/a96e8d5eaa6c9f44a175f09c571f157283cd9ec7/lib/Service/Attachment/AttachmentService.php#L314

## 2.

Our frontend does not pass the attachments when preparing a reply/forward.

Add

```js
attachments: original.attachments,
```

to `data: {...}` all blocks (reply, replyAll and forward) in

https://github.com/nextcloud/mail/blob/a96e8d5eaa6c9f44a175f09c571f157283cd9ec7/src/store/actions.js#L288

**Comment by kesselb** [comment]
Thanks everyone for all the feedback :+1: I think we have more than one issue here :see_no_evil: 

1) The original report and the comment by @boreevyuri for Mail 1.11 (and also 1.12). Patch over here: https://github.com/nextcloud/mail/pull/6543. 
2) As suggested by @sazanof , @miaulalala and @st3iny the issue for Mail 1.12 is that `IMailManager.getMessage` lookup a message by the database id but get's a message uid (which is called messageId to confuse everyone :rofl: ): Patch over here: https://github.com/nextcloud/mail/pull/6544 but it's still a draft as I'm unhappy with mixing our database ids with imap ids when forwarding a message.

**Comment by kesselb** [comment]
Fixed with v1.12.6 / v1.13.4
